### DISCUSSION TITLE
try execpt in @app.middleware("http") is not catching raised HTTPException

### DISCUSSION THREAD

**Discussion by Ismael-Oliv** [fastapi/fastapi#11344] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Union
from fastapi.middleware.cors import CORSMiddleware
from fastapi import FastAPI, Response, Request, HTTPException

app = FastAPI()

origins = ["*"]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_methods=["*"],
    allow_headers=["*"],
)


@app.middleware("http")
async def db_session_middleware(request: Request, call_next):
    response = Response("Internal server error", status_code=500)
    try:
        print("Process started")
        response = await call_next(request)
        print("Not raised exption")

    except Exception as e:
        print("Exceptions raised", e)

    return response


@app.get("/")
def read_root():

    raise HTTPException(status_code=400, detail="Bad request")

    return {"Hello": "World"}


@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    return {"item_id": item_id, "q": q}
```

### Description

Open the browser and access localhost:8000 , it'll show a json with the following 
{
	"detail": "Bad request"
}

but if you go to the terminal that is running fasapi server it didn't print the exception as the code supposed to do in the line print("Exceptions raised", e)

### Operating System

Linux

### FastAPI Version

 0.110.0

### Pydantic Version

 2.6.4

### Python Version

3.10.12

**Comment by YuriiMotov** [comment, marked as answer]
That's because FastAPI (Starlette) wraps all you endpoints by `ExceptionMiddleware` that catches all the exceptions in the endpoint function and dependencies, and calls exception handlers (https://fastapi.tiangolo.com/tutorial/handling-errors/?h=#install-custom-exception-handlers).
But if exception happens inside another middleware (which is closer to endpoint function), you can catch it in your middleware: 

```
from typing import Union

from fastapi import FastAPI, HTTPException, Request, Response
from fastapi.middleware.cors import CORSMiddleware

app = FastAPI()

origins = ["*"]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_methods=["*"],
    allow_headers=["*"],
)

@app.middleware("http")
async def raise_exception_middleware(request: Request, call_next):
    return ("ddd" / 10)

@app.middleware("http")
async def db_session_middleware(request: Request, call_next):
    response = Response("Internal server error", status_code=500)
    try:
        print("Process started")
        response = await call_next(request)
        print("Not raised exption")

    except Exception as e:
        print("Exceptions raised", e)

    return response

@app.get("/")
def read_root():
    raise HTTPException(status_code=400, detail="Bad request")
    return {"Hello": "World"}

@app.get("/items/{item_id}")
def read_item(item_id: int, q: Union[str, None] = None):
    return {"item_id": item_id, "q": q}
```

And if you don't catch it, the `ServerErrorMiddleware` will catch it and return 500 response.
https://www.starlette.io/middleware/

**Reply by Ismael-Oliv** [reply]
I undestood my friend, thank you so  much. But if I want to rollback everything that  it has been done in the database and the middleware (@app.middleware("http")) don't catch it. What I supposed to do? Could you help me please?

**Reply by Ismael-Oliv** [reply]
I got it.

I put the rollback in the:

@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(request, exc):
    request.state.db.rollback()
    return JSONResponse({"error": str(exc.detail)}, status_code=exc.status_code)

**Reply by YuriiMotov** [reply]
You can also decide whether you need to rollback or not by reading the `response.status_code`.
But actually it looks not good, IMHO. Better way is just to use session without autocommit. Tat will perform rollback automatically if you don't do commit or if error occurs during the commit.

**Reply by Ismael-Oliv** [reply]
I've done this way, What do you think?

```
SessionLocal = sessionmaker(autocommit= False, autoflush=False, bind= engine)

@app.exception_handler(StarletteHTTPException)
async def http_exception_handler(request, exc):
    request.state.db.rollback()
    return JSONResponse({"error": str(exc.detail)}, status_code=exc.status_code)


@app.middleware("http")
async def db_session_middleware(request: Request, call_next):
    response = Response("Internal server error", status_code=500)
    try:
        request.state.db = SessionLocal()
        request.state.db.begin()

        response = await call_next(request)
        request.state.db.commit()

    except Exception as E:
        print("Exception from middleware", E)
        request.state.db.rollback()

    finally:
        request.state.db.close()

    return response
```

When its raised a HTTPException, it goes to http_exception_handler it rollsback everything and then it try to commit but there's nothing to commit... cause if another type of Exception is raised the try except catch it.

**Reply by YuriiMotov** [reply]
I don't like this approach.
You spread commit\rollback logic across 2 functions. But you (or you colleges) may forgot and also call commit inside the endpoint function.
I prefer calling `commit` explicitly inside the endpoint function.

And I also think, that it's better to inject session using dependency. In this case you can easily override it for tests.

**Reply by Ismael-Oliv** [reply]
Ok I'll consider it ... Thank you so much for the help my friend.
